File: src/routes/services/prediction.ts
/**
 * 预测服务
 */

import { invokeCommand } from './core';
import type {
  TrainingRequest,
  TrainingResult,
  PredictionRequest,
  PredictionResponse,
  TechnicalOnlyRequest,
  ModelInfo,
  EvaluationResult,
  BacktestRequest,
  BacktestReport,
  MultiTimeframeSignal,
  ProfessionalPredictionResponse,
  OptimizationSuggestions,
} from '../types';

// =============================================================================
// 模型管理
// =============================================================================

/**
 * 列出模型
 */
export async function listModels(symbol: string): Promise<ModelInfo[]> {
  return invokeCommand<ModelInfo[]>('list_stock_prediction_models', { symbol });
}

/**
 * 删除模型
 */
export async function deleteModel(modelId: string): Promise<void> {
  await invokeCommand('delete_stock_prediction_model', { modelId });
}

// =============================================================================
// 训练
// =============================================================================

/**
 * 训练模型
 */
export async function trainModel(request: TrainingRequest): Promise<TrainingResult> {
  return invokeCommand<TrainingResult>('train_stock_prediction_model', { request });
}

/**
 * 使用 Candle 训练模型
 */
export async function trainCandleModel(request: TrainingRequest): Promise<TrainingResult> {
  return invokeCommand<TrainingResult>('train_candle_model', { request });
}

/**
 * 重新训练模型
 */
export async function retrainModel(
  modelId: string,
  epochs: number,
  batchSize: number,
  learningRate: number
): Promise<void> {
  await invokeCommand('retrain_candle_model', {
    modelId,
    epochs,
    batchSize,
    learningRate,
  });
}

// =============================================================================
// 预测
// =============================================================================

/**
 * 预测股价
 */
export async function predictStockPrice(request: PredictionRequest): Promise<PredictionResponse> {
  return invokeCommand<PredictionResponse>('predict_stock_price', { request });
}

/**
 * 使用 Candle 预测
 */
export async function predictWithCandle(request: PredictionRequest): Promise<PredictionResponse> {
  return invokeCommand<PredictionResponse>('predict_with_candle', { request });
}

/**
 * 简化策略预测
 */
export async function predictSimple(request: PredictionRequest): Promise<PredictionResponse> {
  return invokeCommand<PredictionResponse>('predict_candle_price_simple', { request });
}

/**
 * 专业策略预测
 */
export async function predictProfessional(
  request: PredictionRequest
): Promise<ProfessionalPredictionResponse> {
  return invokeCommand<ProfessionalPredictionResponse>('predict_with_professional_strategy', { request });
}

/**
 * 纯技术分析预测
 */
export async function predictTechnicalOnly(
  request: TechnicalOnlyRequest
): Promise<ProfessionalPredictionResponse> {
  return invokeCommand<ProfessionalPredictionResponse>('predict_with_technical_only', { request });
}

// =============================================================================
// 评估与回测
// =============================================================================

/**
 * 评估模型
 */
export async function evaluateModel(modelId: string): Promise<EvaluationResult> {
  return invokeCommand<EvaluationResult>('evaluate_candle_model', { modelId });
}

/**
 * 执行回测
 */
export async function runBacktest(request: BacktestRequest): Promise<BacktestReport> {
  return invokeCommand<BacktestReport>('run_model_backtest', { request });
}

/**
 * 获取优化建议
 */
export async function getOptimizationSuggestions(
  stockCode: string,
  modelName: string,
  backtestReport: BacktestReport
): Promise<OptimizationSuggestions> {
  return invokeCommand<OptimizationSuggestions>('get_optimization_suggestions', {
    stockCode,
    modelName,
    backtestReport,
  });
}

// =============================================================================
// 多周期分析
// =============================================================================

/**
 * 获取多周期信号
 */
export async function getMultiTimeframeSignals(symbol: string): Promise<MultiTimeframeSignal[]> {
  return invokeCommand<MultiTimeframeSignal[]>('get_multi_timeframe_signals', { symbol });
}

/**
 * 获取最新多周期信号
 */
export async function getLatestMultiTimeframeSignal(
  symbol: string
): Promise<MultiTimeframeSignal | null> {
  return invokeCommand<MultiTimeframeSignal | null>('get_latest_multi_timeframe_signal', { symbol });
}

/**
 * 分析多周期预测价值
 */
export async function analyzeMultiTimeframePredictionValue(
  symbol: string
): Promise<Record<string, number>> {
  return invokeCommand<Record<string, number>>('analyze_multi_timeframe_prediction_value', { symbol });
}

File: src-tauri/src/services/prediction.rs
//! 预测服务

use crate::prediction::{
    types::*,
    model::{training, inference, management},
    strategy::multi_timeframe,
};
use crate::db::{connection::create_temp_pool, repository::get_recent_historical_data};

/// 训练模型
pub async fn train_model(request: TrainingRequest) -> Result<TrainingResult, String> {
    training::train_model(request).await
}

/// 重训练模型
pub async fn retrain_model(
    model_id: String,
    epochs: u32,
    batch_size: u32,
    learning_rate: f64,
) -> Result<(), String> {
    training::retrain_model(model_id, epochs, batch_size, learning_rate).await
}

/// 进行预测
pub async fn predict(request: PredictionRequest) -> Result<PredictionResponse, String> {
    inference::predict(request).await
}

/// 简化策略预测
pub async fn predict_simple(request: PredictionRequest) -> Result<PredictionResponse, String> {
    inference::predict_simple(request).await
}

/// 评估模型
pub async fn evaluate_model(model_id: String) -> Result<EvaluationResult, String> {
    inference::evaluate_model(model_id).await
}

/// 列出模型
pub fn list_models(stock_code: &str) -> Vec<ModelInfo> {
    management::list_models(stock_code)
}

/// 删除模型
pub fn delete_model(model_id: &str) -> Result<(), String> {
    management::delete_model(model_id)
}

/// 获取多周期信号
pub async fn get_multi_timeframe_signals(symbol: String) -> Result<Vec<multi_timeframe::MultiTimeframeSignal>, String> {
    let pool = create_temp_pool().await?;
    let historical = get_recent_historical_data(&symbol, 200, &pool)
        .await
        .map_err(|e| format!("获取历史数据失败: {e}"))?;
    
    if historical.is_empty() {
        return Err("未找到历史数据".to_string());
    }
    
    let prices: Vec<f64> = historical.iter().map(|h| h.close).collect();
    let highs: Vec<f64> = historical.iter().map(|h| h.high).collect();
    let lows: Vec<f64> = historical.iter().map(|h| h.low).collect();
    let dates: Vec<String> = historical.iter().map(|h| h.date.format("%Y-%m-%d").to_string()).collect();
    
    let signals = multi_timeframe::generate_multi_timeframe_signals(&prices, &highs, &lows, &dates);
    
    Ok(signals)
}

/// 获取最新多周期信号
pub async fn get_latest_multi_timeframe_signal(symbol: String) -> Result<Option<multi_timeframe::MultiTimeframeSignal>, String> {
    let pool = create_temp_pool().await?;
    let historical = get_recent_historical_data(&symbol, 60, &pool)
        .await
        .map_err(|e| format!("获取历史数据失败: {e}"))?;
    
    if historical.is_empty() {
        return Err("未找到历史数据".to_string());
    }
    
    let prices: Vec<f64> = historical.iter().map(|h| h.close).collect();
    let highs: Vec<f64> = historical.iter().map(|h| h.high).collect();
    let lows: Vec<f64> = historical.iter().map(|h| h.low).collect();
    let date = historical.last().unwrap().date.format("%Y-%m-%d").to_string();
    
    let signal = multi_timeframe::get_latest_signal(&prices, &highs, &lows, &date);
    
    Ok(signal)
}

